include/lbann/layers/learning/convolution.hpp

```diff
@@ -159,8 +193,134 @@ class convolution_layer : public base_convolution_layer<TensorDataType, Device>
     }
   }
 
+#ifdef LBANN_HAS_DISTCONV
+  friend class convolution_distconv_adapter<TensorDataType, Layout, Device>;
+ protected:
+  void setup_distconv_adapter() override {
+    this->get_distconv_adapter_ptr() = make_unique<
+      convolution_distconv_adapter<TensorDataType, Layout, Device>>(*this);
+  }
+
+  bool is_distconv_supported() const override {
+    const auto& kernel_dims = get_kernel_dims();
+    for(int i = 0; i < dc::get_num_spatial_dims(*this); i++) {
+      if (kernel_dims[2 + i] != kernel_dims[2]) {
+        dc::MPIRootPrintStreamDebug()
+            << "Nonsymmetric kernel not supported";
+        return false;
+      }
+      if (kernel_dims[2 + i] !=
+          this->m_pads[i] / this->m_dilations[i] * 2 + 1) {
+        dc::MPIRootPrintStreamDebug()
+            << "Unsupported as padding does not match the kernel size";
+        return false;
+      }
+    }
+    return true;
+  }
+#endif // LBANN_HAS_DISTCONV
 };
 
+#ifdef LBANN_HAS_DISTCONV
+template <typename TensorDataType, data_layout T_layout, El::Device Dev>
+void convolution_distconv_adapter<TensorDataType, T_layout, Dev>::
+setup_distributions(tensor_overlap_constraints &constraints) {
+  base_convolution_adapter<TensorDataType, Dev>::setup_distributions(
+      constraints);
+  auto &l = dynamic_cast<convolution_layer<
+    TensorDataType, T_layout, Dev>&>(this->layer());
+  auto kernel_dims = l.get_kernel_dims();
+  std::reverse(kernel_dims.begin(), kernel_dims.end());
```

Comment:
For clarification, are distconv tensor dimensions in small-stride-to-large-stride order? So data tensors are WHCN and conv kernels are WHCK?

Reply:
That is correct. It's column major.